### PR TITLE
Add PhpCodeValidityTest

### PR DESCRIPTION
**Pull request by Al2Klimov**

### CODE REVIEW

**Comment by Al2Klimov** [comment]
It helped discovering e.g. #3306.

**Comment by Al2Klimov** [comment]
... and cf85e66

**Comment by Thomas-Gelf** [comment]
Cool, makes sense. Please try to explain this briefly in the commit message. I'll add some more comments inline.

File: modules/test/application/clicommands/PhpCommand.php [hunk 1]
```diff
@@ -155,6 +158,96 @@ public function styleAction()
...
+            T_CLASS     => 'class',
+            T_INTERFACE => 'interface'
+        );
+
+        if (version_compare(PHP_VERSION, '5.4.0') > -1) {
+            $types[T_TRAIT] = 'trait';
+        }
+
+        $files = array();
+        $baseDir = realpath(__DIR__ . '/../../../..');
```
Comment (Thomas-Gelf) [inline modules/test/application/clicommands/PhpCommand.php]: Is there any reason not to base this on Icinga\Application methods?

File: modules/test/application/clicommands/PhpCommand.php [hunk 2]
```diff
@@ -155,6 +158,96 @@ public function styleAction()
...
+        foreach ($iterator as $path => $info) {
+            /** @var \SplFileInfo $info */
+            if (preg_match(
+                '~\A(?:test|vendor|modules/[^/]+/test|library/Icinga/Test)(?:/|\z)~',
+                $iterator->getInnerIterator()->getSubPath()
+            ) || ! ($info->isFile() && preg_match('/\.php\z/', $path))) {
+                continue;
+            }
+
+            $content = file_get_contents("file://$path");
```
Comment (Thomas-Gelf) [inline modules/test/application/clicommands/PhpCommand.php]: Do we have to expect a $path to not be a file, therefor asking for an explicit `file://` to protect us?

Comment (Al2Klimov) [reply modules/test/application/clicommands/PhpCommand.php]: I did that on principle.

File: modules/test/application/clicommands/PhpCommand.php [hunk 3]
```diff
@@ -155,6 +158,96 @@ public function styleAction()
...
+    /**
+     * Run code-validity checks
+     *
+     * This command checks whether icingaweb and installed modules match PHP syntax.
+     *
+     * USAGE
+     *
+     * icingacli test php validity
+     */
+    public function validityAction()
```
Comment (Thomas-Gelf) [inline modules/test/application/clicommands/PhpCommand.php]: A CLI Action is like a web Controller, at best it shouldn't contain any business logic. Could you please split this method into multiple ones and move it to a correspondingly named class, maybe in the Icinga\Test namespace?

File: modules/test/application/clicommands/PhpCommand.php (hunk 2 above)
Comment (Thomas-Gelf) [reply modules/test/application/clicommands/PhpCommand.php]: Ok, then please remove it - on principle :laughing:. It absolutely makes sense when opening files based on foreign input, but it is useless and unexpected here.

**Comment by Al2Klimov** [comment]
@Thomas-Gelf Do you agree w/ me on
1. this fails only on PHP 5.3
2. it's a bug in PHP 5.3
3. It's time for #3213

**Comment by Thomas-Gelf** [comment]
Please keep discussion related to #3213 on that ticket, I added a comment there. What exactly breaks on 5.3 here in this issue?

**Comment by Al2Klimov** [comment]
https://travis-ci.org/Icinga/icingaweb2/jobs/332709824#L347

**Comment by Thomas-Gelf** [comment]
That's a bad one, I've met this friend before :laughing:. As you probably already figured out it is triggered by a class with the same name existing in the same namespace. It could be fixed by aliasing the import I guess.

**Comment by Thomas-Gelf** [comment]
(and yes, it's a bug in 5.3)

**Comment by Al2Klimov** [comment]
Damn. Now I understand... and provide an alternative to #3213: https://github.com/PhenX/php-svg-lib/pull/28

**Comment by Thomas-Gelf** [comment]
The linked fix looks good to me. I guess `use Svg\Gradient\Stop as GradiendStop;` would also have fixed this, in case you want to see the full import in the `use` list.

**Comment by Al2Klimov** [comment]
Should we patch our shipped version until we update the lib "one nice day" and the change stays the same (but then it comes from upstream)?

**Comment by Thomas-Gelf** [comment]
Patch it. We should try to avoid doing so, but as we already modified this library before - go for it.

File: modules/test/application/clicommands/PhpCommand.php (hunk 3 above)
Comment (Al2Klimov) [reply modules/test/application/clicommands/PhpCommand.php]: What for? No one else uses that "business logic". (And I hope no one will.)

**Comment by Al2Klimov** [comment]
Should we? Aren't our components' tests just for our components?

**Comment by nilmerg** [comment]
No-one prevents me from setting just my own module path. Then also not only our own components are covered but mine as well. And since setting multiple module paths is generally supported it should be here as well.

At least in my dev environment the tests fail because the audit module is not loaded due to it being installed in a separate path.

File: modules/test/application/clicommands/PhpCommand.php (hunk 1 above)
Comment (Al2Klimov) [reply modules/test/application/clicommands/PhpCommand.php]: Yes: This is PhpUnit, not our fully-bootstrapped CLI.

**Comment by Al2Klimov** [comment]
TODO:

* [ ] #3771

**Comment by nilmerg** [comment]
Still doesn't work even with #3852.

**Comment by nilmerg** [comment]
There are lint tests in the actions now.

**Pull request closed without merging**